### PR TITLE
added rational-modeline.el

### LINKED ISSUE #44: make doom-modeline optional
im playing around with [powerline](https://github.com/milkypostman/powerline/tree/master) and similar packages and as far as i understand it is much closer to vanilla emacs modeline while allowing for lots of customization

im also thinking about playing around with [telephone-line](https://github.com/dbordak/telephone-line/tree/master) which seems to do ruffly the same

### PR DESCRIPTION
**Pull request by erikLundstedt**
this closes #44 and replaces #45 

moved existing configs to the new module
and added 2 lines to example file to show how to use it

Signed-off-by: erik.lundstedt <erik@lundstedt.it>

### CODE REVIEW

File: modules/rational-modeline.el [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+;; Commentary
+
+;;
+
+;;; Code:
+
+(defcustom rational-modeline/type "doom"
+  "one of 'GNU' 'doom' 'telephone' ";;'space WIP' "
+  )
```
Comment (jeastman) [inline modules/rational-modeline.el:16]: You will need to handle the situation where this changes (e.g. with `:set`), otherwise it will not take effect unless it is configured _before_ the module is required.

`rational-modeline/type` should be `rational-modeline-type` for consistency with style guide.

File: modules/rational-modeline.el [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  ;; (defun test ()            ;;
+  ;;   (message "hello world") ;;
+  ;;   )                       ;;
+  ;;                           ;;
+  ;;   (setq a 'test)          ;;
+  ;;                           ;;
+  ;;                           ;;
+  ;;                           ;;
+  ;; (funcall a)               ;;
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (jeastman) [inline modules/rational-modeline.el:116]: Probably should remove the test comment block

File: modules/rational-ui.el [hunk 3]
```diff
@@ -13,7 +13,7 @@
 ;;; Code:
 
 (straight-use-package 'all-the-icons)
-(straight-use-package 'doom-modeline)
+;;(straight-use-package 'doom-modeline)
```
Comment (jeastman) [inline modules/rational-ui.el:16]: No need to leave the comment in file here

File: modules/rational-ui.el [hunk 4]
```diff
@@ -27,16 +27,18 @@ Use a plist with the same key names as accepted by `set-face-attribute'.")
...
+;;                                                                                         ;;
+;; ;; Start up the modeline after initialization is finished                               ;;
+;; (add-hook 'after-init-hook 'doom-modeline-init)                                         ;;
+;;                                                                                         ;;
+;; ;; Configure `doom-modeline'                                                            ;;
+;; (customize-set-variable 'doom-modeline-height 15)                                       ;;
+;; (customize-set-variable 'doom-modeline-bar-width 6)                                     ;;
+;; (customize-set-variable 'doom-modeline-minor-modes t)                                   ;;
+;; (customize-set-variable 'doom-modeline-buffer-file-name-style 'truncate-except-project) ;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (jeastman) [inline modules/rational-ui.el:41]: No need to leave comment in file here.

File: modules/rational-modeline.el [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+;;
+
+;;; Code:
+
+(defcustom rational-modeline/type "doom"
+  "one of 'GNU' 'doom' 'telephone' ";;'space WIP' "
+  )
+
+
+(when (string-equal rational-modeline/type "GNU")
```
Comment (jeastman) [inline modules/rational-modeline.el:19]: Each `when` would need to be handled when the custom variable changes.

File: modules/rational-modeline.el [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  ;;TODO make it look as good as possible without external packages
+  )
+
+
+
+
+(when (string-equal rational-modeline/type "doom")
```
Comment (jeastman) [inline modules/rational-modeline.el:26]: Each `when` would need to be handled when the custom variable changes.

File: modules/rational-modeline.el [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  (customize-set-variable 'doom-modeline-height 15)
+  (customize-set-variable 'doom-modeline-minor-modes t)
+  )
+
+
+
+
+
+
+(when (string-equal rational-modeline/type "telephone")
```
Comment (jeastman) [inline modules/rational-modeline.el:40]: Each `when` would need to be handled when the custom variable changes.

File: modules/rational-modeline.el [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    )
+
+
+
+
+
+  (straight-use-package 'telephone-line)
+
+
+  (defcustom rational-modeline/telephoneline-style 'telephoneline-examples-cubed
```
Comment (jeastman) [inline modules/rational-modeline.el:101]: Also want a `:set` here to handle when the value changes.

**Review by jeastman — CHANGES_REQUESTED** [review]

File: modules/rational-modeline.el (hunk 1 above)
Comment (erikLundstedt) [reply modules/rational-modeline.el:16]: > You will need to handle the situation where this changes (e.g. with :set)
not sure how to do this, happy to get an example or something like that.

I havent really written a lot of elisp code and allmost none outside of my personal configs where this doesnt matter too much

Comment (erikLundstedt) [reply modules/rational-modeline.el:17]: > rational-modeline/type should be rational-modeline-type for consistency with style guide.
this should be fixed now



not sure how to add changes to pr... but i fixed this in a recent push

File: modules/rational-modeline.el (hunk 5 above)
Comment (erikLundstedt) [reply modules/rational-modeline.el]: How would I do this?

I am pretty new to elisp and would be glad for any help you can give

File: modules/rational-modeline.el (hunk 1 above)
Comment (jeastman) [reply modules/rational-modeline.el:16]: > I havent really written a lot of elisp code and allmost none outside of my personal configs where this doesnt matter too much

No worries. One of the Principles of Rational Emacs: "Helps you learn Emacs Lisp"

File: modules/rational-modeline.el (hunk 5 above)
Comment (jeastman) [reply modules/rational-modeline.el]: Check out the implementation of `rational-ui-display-line-numbers` in rational-ui.el. Notice the `:set` on the custom variable and how it responds to the variable being changed. Just make sure you use `set-default` in the body of your `:set` to ensure the value actually gets changed. You can test by setting the value (with `custom-set-variables`). Try setting the value both before and after `require`ing `rational-modeline` to see the effect.  Happy to help if you have questions after looking at `rational-ui-display-line-numbers`.

File: modules/rational-modeline.el [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+;;; Code:
+
+(defcustom test 'nil
+  "something here"
+  :group 'rational
+  :set (lambda (sym val)
+         (set-default sym val)
+         (when (equal test t)
+           (counsel-find-file)
```
Comment (bkaestner) [inline modules/rational-modeline.el:20]: This shouldn't be in a productive pull request. Please remove the test code.

File: modules/rational-modeline.el [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+;;(when (string-equal rational-modeline-type "GNU")
+;;TODO make it look as good as possible without external packages
+;;  )
+
+
+
+
+(defun doom-modeline-type ()
+  (straight-use-package 'doom-modeline)
```
Comment (bkaestner) [inline modules/rational-modeline.el:34]: None of these functions have either the `rational-` prefix, nor a docstring. Please add some documentation.

File: modules/rational-modeline.el [hunk 11]
```diff
@@ -0,0 +1,158 @@
+;;;; rational-modeline.el --- rational module for setting up the modeline  -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2022
+;; SPDX-License-Identifier: MIT
+
+;; Authors: System Crafters Community, David "System Crafters" Wilson, Erik Lundstedt
```
Comment (bkaestner) [inline modules/rational-modeline.el:6]: I don't think I have heard of @daviwil call himself _David "System Crafters" Wilson_ 🤔

File: custom-vars.el [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+ '(scroll-conservatively 101)
+ '(scroll-margin 0)
+ '(scroll-preserve-screen-position t)
+ '(visible-bell 1))
+(custom-set-faces
+ ;; custom-set-faces was added by Custom.
+ ;; If you edit it by hand, you could mess it up, so be careful.
+ ;; Your init file should contain only one such instance.
+ ;; If there is more than one, they won't work right.
+ )
```
Comment (bkaestner) [inline custom-vars.el:39]: This whole file seems like an accidental commit. Was this really intended? Otherwise please rebase to remove the file.

**Review by bkaestner — CHANGES_REQUESTED** [review]

File: custom-vars.el (hunk 12 above)
Comment (erikLundstedt) [reply custom-vars.el:39]: Yeah that was accidental

**Comment by erikLundstedt** [comment]
I am thinking that if the user wants to use something not supported by the module, they can just not require it, at the moment I'm using this branch+with the setting for telephone-line which works great for me and seems to be pretty much what you wanted (according to what you(@daviwil ) said about it in the video) by default the module uses doom modeline though as I don't want to break configs, I also added the example configs for telephone-line for ease of use but the user is able to write their own if they want

Telephone-line seems to do for the modeline what vertico does for completion, it makes it look good while keeping it simple, lightweight and vanilla-like

File: custom-vars.el (hunk 12 above)
Comment (jeastman) [reply custom-vars.el:39]: You will need to remove the file from your PR as it does not belong.

File: history [hunk 13]
```diff
@@ -0,0 +1,16 @@
+;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
```
Comment (jeastman) [inline history:1]: This file also does not appear to be related to the commit. Another accidental inclusion?

File: modules/rational-modeline.el (hunk 11 above)
Comment (jeastman) [reply modules/rational-modeline.el:6]: Yes, should probably update the comment.

File: modules/rational-modeline.el (hunk 9 above)
Comment (jeastman) [reply modules/rational-modeline.el:20]: Agree this should also be removed.

**Review by jeastman — CHANGES_REQUESTED** [review]

File: modules/rational-modeline.el (hunk 1 above)
Comment (jeffbowman) [reply modules/rational-modeline.el:23]: For the choice of options, you should probably use something like the following in your `defcusom` definition:

``` emacs-lisp
    :type '(choice (const :tag "GNU default" gnu)
                   (const :tag "Doom modeline" doom)
                   (const :tag "Telephone modeline" telephone))
```

**Comment by erikLundstedt** [comment]
im closing this as i moved the code to my own config folder
